### PR TITLE
Migrate Result and ChatError to Stream Result dependency

### PR DESCRIPTION
**Pull request by skydoves**
### 🎯 Goal

Standardize Stream Chat's Result/ChatError class into an independent package and re-use it for multiple use cases. Video SDK also needs similar results & error-handling models, and it must be very confusing if our clients use them in the same place. We can provide better developer experiences by extracting the same design model into a common package ([stream-result-draft](https://github.com/GetStream/stream-result-draft)).

### 🛠 Implementation details

- `io.getstream.chat.android.client.utils.Result` was migrated to `io.getstream.result.Result` package.
- `io.getstream.chat.android.client.errors.ChatError` was migrated to `io.getstream.result.StreamError` package.
- `result.chatErrorOrNull()` was renamed with  `result.streamErrorOrNull()`.

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch

#### Code & documentation
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

gif

### CODE REVIEW

File: docusaurus/docs/Android/01-basics/05-core-concepts.mdx [hunk 1]
```diff
@@ -111,7 +111,7 @@ if (result.isSuccess()) {
     Channel channel = result.getOrNull();
 } else {
     // Handle error
-    ChatError error = result.chatErrorOrNull();
+    ChatError error = result.streamErrorOrNull();
```
Comment (kanat) [inline docusaurus/docs/Android/01-basics/05-core-concepts.mdx:114]:
```suggestion
    StreamError error = result.streamErrorOrNull();
```

File: stream-chat-android-client/src/androidTest/java/io/getstream/chat/android/client/utils/TestInitCallback.kt [hunk 2]
```diff
@@ -34,9 +35,9 @@ internal class TestInitCallback : Call.Callback<ConnectionData> {
 
     override fun onResult(result: Result<ConnectionData>) {
         if (result.isSuccess) {
-            data = result.data()
+            data = result.getOrThrow()
         } else {
-            error = result.error()
+            error = result.streamErrorOrNull()
```
Comment (kanat) [inline stream-chat-android-client/src/androidTest/java/io/getstream/chat/android/client/utils/TestInitCallback.kt:40]:
```suggestion
            error = result.streamErrorOrThrow()
```
I assume this is supposed to be something like `result.streamErrorOrThrow()` like `error()` before. WDYT?

```kotlin
    public fun error(): ChatError {
        return checkNotNull(error) {
            "Result is successful, not an error. Check result.isSuccess before reading the error."
        }
    }
```

**Review by kanat — APPROVED** [review]
Overall, LGTM

File: docusaurus/docs/Android/01-basics/05-core-concepts.mdx (hunk 1 above)
Comment (skydoves) [reply docusaurus/docs/Android/01-basics/05-core-concepts.mdx]: done: [57f8687](https://github.com/GetStream/stream-chat-android/pull/4755/commits/57f868731206eb75f1c9c94b451a8866898046e8)

**Comment by skydoves** [comment]
Will be updated to use the stable version: https://github.com/GetStream/stream-result

File: stream-chat-android-client/build.gradle [hunk 3]
```diff
@@ -86,6 +86,7 @@ dependencies {
     // Tests
     testImplementation project(':stream-chat-android-test')
     testImplementation project(":stream-chat-android-client-test")
+    testImplementation Dependencies.streamResult
```
[the remark this answers is not in the document]
Comment (skydoves) [reply stream-chat-android-client/build.gradle:89]: We're also adding on the `stream-chat-android-offline` module.

**Pull request merged**